### PR TITLE
[RNMobile] Refactor native gradient logic

### PR DESCRIPTION
**Pull request by lukewalczak**
## Description

Refactor the parsing gradient which is passed to rn linear gradient component.

## How has this been tested?

1. Set theme `Twenty Twenty`
2. Open a post
3. Add cover block
4. Choose the background color
5. Press cog to open settings
6. Go to color and press `Gradient`
**Expect there is no crash**
7. Choose one of the gradients
8. Save the post

9. Exit the editor and switch theme to `Spearhead`
10. Kill the app and reopen it (known issue with themes)
11. Go to the previous post
**Expect cover background is not set**
12. Choose the background color
13. Press cog to open settings
14. Go to color and press `Gradient`

**Expect there is no crash and expect to see the following gradients**:
image

15. Choose one of them
16. Save the post
17. Open the post on web and compare if gradients looks alike

## TO INVESTIGATE

Once trying to convert marked gradients to radial gradient (press `customize gradient` then `radial`) app crashes. I've checked that on the web and there user cannot convert these two gradients to linear. 

Screenshot 2020-11-26 at 17 53 16

I push the commit with the fix and left the [comment](https://github.com/WordPress/gutenberg/pull/27307/files#r531164626)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/custom-gradient-picker/serializer.js
Comment (lukewalczak) [inline packages/components/src/custom-gradient-picker/serializer.js:24]: Gradient color `length` _(gradient color location - e.g. 50%)_ can be undefined e.g. in `linear-gradient(to top , #3C8067, #FAFBF6)` from the theme called `Spearhead`. On web nothing happens when trying to switch from `linear` to `radial` but on mobile there is a crash since we reuse logic within `serializer.js`. 

Properties of undefined `length`  are then used in function:

```
export function serializeGradientPosition( { type, value } ) {
	return `${ value }${ type }`;
}
```
which makes a crash.

**Comment by cameronvoell** [comment]
Hi @ajitbohra @chrisvanpatten 👋 , a review on this update to `packages/components/src/custom-gradient-picker/serializer.js` ([see comment](https://github.com/WordPress/gutenberg/pull/27307/files#r531164626)) from the web perspective would be appreciated. The change is useful for fixing a crash in native mobile apps. I can review the rest of the PR from the mobile perspective. 🙇

**Comment by cameronvoell** [comment]
Tested successfully from a mobile perspective 👍 . I see the change fixed the crash when setting the custom gradient on the cover block for the Spearhead Theme. This change is ready to go as long as the web side change checks out with necessary code owners.

**Comment by hypest** [comment]
`End-to-End Tests / Admin - 3 (pull_request) Failing after 16m — Admin - 3` failed so, I went ahead and restarted the job that includes it.

**Comment by hypest** [comment]
The failing job also [fails on master](https://github.com/WordPress/gutenberg/runs/1462490886) so, doesn't seem like this PR is introducing the fail.

File: packages/components/src/custom-gradient-picker/serializer.js
Comment (ellatrix) [reply packages/components/src/custom-gradient-picker/serializer.js:24]: To me this seems ok to do, but this code is new to me as well. Maybe @jorgefilipecosta can have a quick look?

Comment (jorgefilipecosta) [reply packages/components/src/custom-gradient-picker/serializer.js:24]: This change looks good to me 👍 

I guess as an enhancement we can add this logic inside the serializeGradientPosition function directly:
```
export function serializeGradientPosition( position ) {
	if ( ! position ) {
		return '';
	}
	const { value, type } = position;
	return `${ value }${ type }`;
}
```

**Review by jorgefilipecosta — APPROVED** [review]

File: packages/components/src/custom-gradient-picker/serializer.js
Comment (lukewalczak) [reply packages/components/src/custom-gradient-picker/serializer.js:24]: Sure, make sense 👍🏼

**Comment by Tug** [comment]
Just tested on WPiOS, it's indeed fixing the bug reported [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/2830#issuecomment-733963175) 🥳 
Had to fix some compilation error but that should be handled on the WPiOS PR I believe.

**Review by Tug — COMMENTED** [review]
Code looks good but I still have an error when opening the page it was crashing on before.


Also it would be great to have some unit tests for those functions in `mobile/gradient/`.

**Comment by lukewalczak** [comment]
@Tug Is there an option to paste a HTML of that post?

**Comment by lukewalczak** [comment]
> Also it would be great to have some unit tests for those functions in mobile/gradient/.

I will tackle it as a next task I think, that logic definitely need some unit tests.

**Comment by hypest** [comment]
> I will tackle it as a next task I think

+1 for tackling that or similar tasks after we merge and include this fix to the ongoing mobile editor release that's due today.

**Comment by Tug** [comment]
My bad it's working fine, I had trouble with my environment.

**Review by Tug — APPROVED** [review]

**Comment by Tug** [comment]
Looks like https://github.com/WordPress/gutenberg/pull/25854 broke CI tests on `master`. Let's go ahead and merge this one

**Pull request merged**
